src/browser.ts

```diff
@@ -1,41 +1,95 @@
+  return find(paths, p => Boolean(p.deserialize(pathname))) || defaultNoPath;
+}
+
+function locationToPathname(location?: Location): string {
+  return location ? trim(location.pathname, ' /') : '';
+}
+
+function locationToPathConfig(paths: PathConfigs<BrowserState>, location?: Location): PathConfig<BrowserState> {
+  return pathnameToPathConfig(locationToPathname(location), paths);
+}
+
+function parsePathname(paths: PathConfigs<BrowserState>, location?: Location): Partial<BrowserState> {
+  return locationToPathConfig(paths, location).deserialize(locationToPathname(location))!;
+}
+
 // export is for tests only
 // TODO(gio): `location` here is optional just because (I think)
 // it can be undefined in some tests using `memoryHistory`
-export function parseBrowserState(location?: Location): BrowserState {
-  const view = location ? trim(location.pathname, ' /') : '';
+export function parseBrowserState(paths: PathConfigs<BrowserState>, location?: Location): BrowserState {
+  const pathState = parsePathname(paths, location);
   const s = location ? qs.parse(trim(location.search, ' ?')) : {};
-  return { ...s, view };
+  return { ...s, ...pathState };
+}
+
+function cleanupPathParams(
+  s: BrowserState,
+  currentPath: PathConfig<BrowserState>,
+  nextPath: PathConfig<BrowserState>
+): BrowserState {
+  const currentPathParamKeys = Object.keys(currentPath.pick(s));
+  const nextPathParamKeys = Object.keys(nextPath.pick(s));
+  return omit(s, difference(currentPathParamKeys, nextPathParamKeys));
+}
+
+function browserStateToPathConfig(s: BrowserState, paths: PathConfigs<BrowserState>): PathConfig<BrowserState> {
+  return find(paths, p => p.is(s)) || defaultNoPath;
 }
 
 // export is for tests only
-export function serializeBrowserState(s: BrowserState): Partial<Location> {
-  const pathname = `/${encodeURIComponent(s.view)}`;
-  const search = `?${qs.stringify(omit(s, ['view']))}`; // will encode URI by default
+export function serializeBrowserState(s: BrowserState, paths: PathConfigs<BrowserState>): Partial<Location> {
+  const path = browserStateToPathConfig(s, paths);
+  const pathState = path.pick(s);
+  const pathname = `/${path.serialize(mapValues(pathState, encodeURIComponent))}`;
+  const search = `?${qs.stringify(omit(s, Object.keys(pathState)))}`; // will encode URI by default
   return { pathname, search };
 }
 
-export default function(history: History = createBrowserHistory()) {
+export default function(paths: PathConfigs<BrowserState> = [], history: History = createBrowserHistory()) {
+  let currentPath = locationToPathConfig(paths, history.location);
+
   function syncToBrowser(s: BrowserState, push: boolean = true): void {
-    (push ? history.push : history.replace)(serializeBrowserState(s));
+    (push ? history.push : history.replace)(serializeBrowserState(s, paths));
   }
 
   function onBrowserChange(callback: (s: BrowserState, action: Action) => void): void {
```

Comment:
why a callback instead of a `Promise`?

Reply:
the callback is called multiple times, not just once